### PR TITLE
Fix edge cases around non-thread relations to thread roots and read receipts

### PR DESCRIPTION
**Pull request by t3chguy**
Which caused stuck notifications

~Requires https://github.com/matrix-org/matrix-react-sdk/pull/11294~

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Fix edge cases around non-thread relations to thread roots and read receipts ([\#3607](https://github.com/matrix-org/matrix-js-sdk/pull/3607)).

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -4973,43 +4973,84 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
 
+        let thread: string | null;
+        let body: Record<string, any>;
+
+        if ((bodyOrThread !== null && typeof bodyOrThread === "object") || typeof unthreadedOrBody === "boolean") {
+            // Backwards compatible signature case
+            thread = !unthreadedOrBody ? this.guessThreadForReceipt(event) : null;
+            body = (bodyOrThread as Record<string, any>) ?? {};
+        } else {
+            thread = bodyOrThread as string | null;
```
Comment (andybalaam) [inline src/client.ts:5038]: `|` -> `||` ?

Comment (andybalaam) [reply src/client.ts:5038]: Oh no, ignore me.

**Review by andybalaam — APPROVED** [review]
Looks good, one tiny comment.

File: spec/unit/read-receipt.spec.ts [hunk 2]
```diff
@@ -123,7 +145,91 @@ describe("Read receipt", () => {
...
+                })
+                .respond(200, {});
+
+            client.sendReadReceipt(threadEvent, ReceiptType.Read, null);
+
+            await httpBackend.flushAllExpected();
+            await flushPromises();
+        });
+
+        it("sends an main timeline receipt for a thread root (deprecated)", async () => {
```
Comment (t3chguy) [inline spec/unit/read-receipt.spec.ts:169]:
```suggestion
        it("sends a main timeline receipt for a thread root (deprecated)", async () => {
```

File: spec/unit/read-receipt.spec.ts [hunk 3]
```diff
@@ -123,7 +145,91 @@ describe("Read receipt", () => {
...
+                    expect(request.data.thread_id).toBe("main");
+                })
+                .respond(200, {});
+            client.sendReadReceipt(threadRootEvent, ReceiptType.Read, false);
+
+            await httpBackend.flushAllExpected();
+            await flushPromises();
+        });
+
+        it("sends an main timeline receipt for a reaction to a thread root (deprecated)", async () => {
```
Comment (t3chguy) [inline spec/unit/read-receipt.spec.ts:200]:
```suggestion
        it("sends a main timeline receipt for a reaction to a thread root (deprecated)", async () => {
```

**Review by andybalaam — CHANGES_REQUESTED** [review]
I just tried this out locally, and edits to thread roots are not reflected until I refresh:

image

(Changing to a different room and back did not fix the problem)

Switching back to develop, I see the change immediately.

**Comment by andybalaam** [comment]
> Screen.Recording.2023-07-19.at.11.46.07.mov
> 
> @andybalaam I cannot reproduce that at all, are you sure your build was fine? I'm using https://pr11294--matrix-react-sdk.netlify.app/

On the netlify build I see the same as you, but when I build locally, with only matrix-js-sdk on `t3chguy/stuck-edits-thread-root` I see the behaviour I described.

I note that there is still a `t3chguy/stuck-edits-thread-root` branch on matrix-react-sdk even though the PR is closed: https://github.com/matrix-org/matrix-react-sdk/pull/11291

Could branch matching be fooling us here?

**Comment by t3chguy** [comment]
https://pr11294--matrix-react-sdk.netlify.app/ is a build from https://github.com/matrix-org/matrix-react-sdk/pull/11294 not https://github.com/matrix-org/matrix-react-sdk/pull/11291 - the branch was reused between PRs and has since been deleted. Given https://github.com/matrix-org/matrix-react-sdk/pull/11294 only touches cypress I don't see how branch matching could be doing anything awry here. Critically in the screencap I included above you can see the edit hidden event is in the main timeline, unlike on develop where it'd be in the thread timeline:

image

**Review by andybalaam — APPROVED** [review]
OK, I just rebuilt locally and it worked fine, so I have no idea what was happening before - I would doubt it happened, except the screenshot shows something.

Good that this is going in now, so we've some time to test before the next RC, but it looks sensible to me, thanks!

**Pull request merged**